packages/programs/program/src/keychain.ts

```diff
@@ -0,0 +1,219 @@
+import { AbstractType, field, option, variant, vec } from "@dao-xyz/borsh";
+import { Program } from "./program";
+import {
+	And,
+	BoolQuery,
+	ByteMatchQuery,
+	Documents,
+	MissingField,
+	Or,
+	PutOperation,
+	Query,
+	SearchRequest,
+	Sort,
+	SortDirection,
+	StateFieldQuery,
+	StringMatch
+} from "../../data/document/src";
+import {
+	EncryptedThing,
+	EncryptedSecp256k1Keypair,
+	EncryptedX25519Keypair,
+	EncryptedEd25519Keypair,
+	EncryptedKeypair,
+	DecryptedThing,
+	createLocalEncryptProvider,
+	X25519Keypair,
+	X25519PublicKey,
+	createDecrypterFromKeyResolver,
+	KeyResolverReturnType,
+	Ed25519PublicKey,
+	Secp256k1PublicKey,
+	Identity,
+	EncryptProvide,
+	KeyExchangeOptions,
+	DecryptProvider
+} from "@peerbit/crypto";
+import { compare } from "@peerbit/uint8arrays";
+
+@variant(0)
+class EncryptedExtendedKey {
+	// TODO: The owner of the key determined by wallet address or public sign key?
+	@field({ type: vec("u8") })
+	owner: Uint8Array;
+	// Is this key revoked? HAS TO BE TRUE if recipient is set.
+	@field({ type: "bool" })
+	revoked: boolean;
+	// Provide a single-use key for someone else
+	@field({ type: option(vec("u8")) })
+	recipient?: Uint8Array;
+	// TODO: add the actual key here
+	@field({ type: EncryptedKeypair })
+	keypair:
+		| EncryptedX25519Keypair
+		| EncryptedEd25519Keypair
+		| EncryptedSecp256k1Keypair;
+
+	constructor(parameters: {
+		owner: Uint8Array;
+		keypair:
+			| EncryptedX25519Keypair
+			| EncryptedEd25519Keypair
+			| EncryptedSecp256k1Keypair;
+		revoked?: boolean;
+	}) {
+		this.keypair = parameters.keypair;
+		this.owner = parameters.owner;
+		this.revoked = parameters.revoked ?? false;
+	}
+}
+
+export class KeychainProgram extends Program {
+	@field({ type: Documents<EncryptedExtendedKey> })
+	keys: Documents<EncryptedExtendedKey>;
```

Comment:
When PR https://github.com/dao-xyz/peerbit/pull/215 is merged it should in practice just be possible to just do 

```js
@field({ type: Documents<EncryptedExtendedKey> })
keys: Documents<Ed25519Keypair>; // or something else secret
```


and later just do 

```
sharedKeychain.keys.put(keypair, { encryption : { hash: ` the hash of the secret key you want to use` } })
```

Reply:
or, well it depends if you want to have a a public and private part of the document of course. Then you need the `EncryptedExtendedKey` type

Reply:
Hmmm, but isn't the problem that the `encryption` property on the `put` method encrypts the whole key? How would I then find the public key of someone else that I want to encrypt something for, if the public key is encrypted as well?

Reply:
Also, how can I mark keys as revoked?

Reply:
And I have also been thinking - sometimes, a user I want to encrypt something for has not generated a keypair yet. So at that time of speaking, I want to generate a keypair for him that is only single use.

Reply:
> property on the put method encrypts the whole key? How would I then find the public key of someone else that I want to encrypt something for, if the public key is encrypted as well?

Ah yee for this to work the decryption has to be performed by keys that exist in a lower level keychain. When you use the put method there PublicKeyEnvelope (in the crypto lib) will leak the publickeys of the parties that can decrypt the message, which allows you to figure out if you can decrypt a message.


I think it might be easier to reason about stuff if you would separate "encryption" utilties from this shared-keychain. So this keychain stores keys that are encrypted, but the keys for providing the encryption  for storing these keys for this keychain perhaps needs to originate from a lower level keychain. 

It could in practice also be possible that keys in this keychain are allso encrypted by keys that also exist in this keychain, but then you need to make sure you are loading the database in the right order kind of (which is possible in practice there is a oplog behind the scenes that allows you to link commits)

Reply:
If you want to revoke keys I think you need a revoked flag on the keypair object, or something like that. Else you could in theory also just use the delete function of the document store and assumes keys are only valid if they exist in the keychain